utils/async_utils.py

```diff
@@ -0,0 +1,321 @@
+            if task.done():
+                continue
+
+            task.cancel()
+            tasks.append(task)
+
+        for store in self.children:
+            tasks.append(store.close())
+
+        await asyncio.gather(*tasks, return_exceptions=True)
+
+
+class AsyncABCMeta(ABCMeta):
+    """
+    This metaclass ensures that the ``__ainit__`` method is a coroutine.
+    """
+    def __new__(
+        cls,
+        clsname: str,
+        bases: tuple[type, ...],
+        namespace: dict[str, Any],
+    ) -> AsyncABCMeta:
+        instance = super(AsyncABCMeta, cls).__new__(
+            cls,
+            clsname,
+            bases,
+            namespace,
+        )
+
+        if not asyncio.iscoroutinefunction(instance.__ainit__):  # type: ignore
+            raise TypeError("__ainit__ must be coroutine")
+
+        return instance
+
+
+class AsyncInstanceType(metaclass=AsyncABCMeta):
+    """
+    Parameters
+    ----------
+    *args: Any
+        The arguments to pass to ``__ainit__``.
+    **kwargs: Any
+        The keyword arguments to pass to ``__ainit__``.
+
+    Attributes
+    ----------
+    loop: :class:`asyncio.AbstractEventLoop`
+        The event loop to use for creating tasks and futures.
+    """
+    __slots__: tuple[str, ...] = ("_args", "_kwargs")
+    _args: tuple[Any, ...]
+    _kwargs: dict[str, Any]
+
+    def __new__(
+        cls,
+        *args: Any,
+        **kwargs: Any,
+    ) -> AsyncInstanceType:
+        instance = super().__new__(cls)
+        instance._args = args
+        instance._kwargs = kwargs
+        return instance
+
+    @property
+    def loop(self) -> asyncio.AbstractEventLoop:
+        return asyncio.get_running_loop()
+
+    def __await__(self) -> Generator[Any, None, AsyncInstanceType]:
+        yield from self.__ainit__(*self._args, **self._kwargs).__await__()
+
+        # blep :<
+        return self
+
+    async def __ainit__(
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        ...
+
+
+class AsyncInstance(AsyncInstanceType):
+    """A base class for creating async classes.
+
+    This class is designed to be used as a base class for creating
+    async classes. It provides a ``__await__`` method which calls
+    ``__ainit__`` and returns the instance.
+    """
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        self.__closed = False
+        self._async_class_task_store: Task
+
+    @property
+    def __tasks__(self) -> Task:
+        return self._async_class_task_store
+
+    @property
+    def is_closed(self) -> bool:
+        return self.__closed
+
+    def compose_task(self, *args: Any, **kwargs: Any) -> asyncio.Task[Any]:
+        return self.__tasks__.compose_task(*args, **kwargs)
+
+    def compose_future(self) -> asyncio.Future[Any]:
+        return self.__tasks__.compose_future()
+
+    async def __adel__(self) -> None:
+        pass
+
+    def __await__(self) -> Generator[Any, None, AsyncInstance]:
```

Comment:
```suggestion
    def __await__(self) -> Generator[Any, None, Self]:
```

Return ``typing.Self`` here so the linter knows which subclass it is when awaiting it.

```py
test = await Test()
reveal_type(test)  # type is AsyncInstance -> type is Test
```